source/config/profileUpgrader.py

```diff
@@ -0,0 +1,67 @@
+# -*- coding: UTF-8 -*-
+#A part of NonVisual Desktop Access (NVDA)
+#Copyright (C) 2016 NV Access Limited
+#This file is covered by the GNU General Public License.
+#See the file COPYING for more details.
+
+from logHandler import log
+from configSpec import latestSchemaVersion, confspec
+from configobj import flatten_errors
+from copy import deepcopy
+import profileUpgradeSteps
+
+schemaVersionKey = "schemaVersion"
+
+def upgrade(profile, validator):
+	# when profile is none or empty we can still validate. It should at least have a version set.
+	log.info("Checking schema version of config:\n%s" % profile)

```

Comment:
In reply to your general comment about logging the config values, there is no where that logs the config after the upgrade. However the config file itself servers that purpose. If there is an error during the upgrade then having the starting point should be enough for us to reproduce it.

When there is an error, the items that dont validate are added to the log. Here is an example of the log file when this happens:

> INFO - __main__ (03:03:44):
> Starting NVDA
> INFO - core.main (03:03:45):
> Config dir: C:\work\nvda\source\userConfig
> INFO - config.ConfigManager._loadConfig (03:03:45):
> Loading config: .\userConfig\nvda.ini
> INFO - config.profileUpgrader.upgrade (03:03:45):
> Checking schema version of config:
> {u'schemaVersion': u'1', u'upgrade': {}, u'update': {}, u'general': {u'showWelcomeDialogAtStartup': u'False', u'language': u'en', u'saveConfigurationOnExit': u'True', u'askToExit': u'False', u'playStartAndExitSounds': u'True', u'loggingLevel': u'DEBUG'}, u'speech': {u'synth': u'espeak', u'espeak': {u'voice': u'en', u'variant': u'max', u'rate': u'30', u'pitch': u'40', u'inflection': u'75', u'volume': u'20'}}, u'braille': {u'cursorBlinkRate': u'-1', u'noBraille': {u'port': u''}}, u'keyboard': {u'useCapsLockAsNVDAModifierKey': u'True'}, u'documentFormatting': {u'autoLanguageSwitching': u'True'}, u'mouse': {u'enableMouseTracking': u'False', u'audioCoordinatesOnMouseMove': u'False', u'audioCoordinates_detectBrightness': u'False'}, u'speechViewer': {u'showSpeechViewerAtStartup': u'True', u'x': u'68', u'y': u'44', u'width': u'500', u'height': u'500', u'displays': [u'(1920, 1080)'], u'autoPositionWindow': u'False'}}
> ERROR - config.ConfigManager._initBaseConf (03:03:45):
> Error loading base configuration
> Traceback (most recent call last):
>   File "config\__init__.py", line 372, in _initBaseConf
>     profile = self._loadConfig(fn) # a blank file will be created if fn does not exist
>   File "config\__init__.py", line 400, in _loadConfig
>     profileUpgrader.upgrade(profile, self.validator)
>   File "config\profileUpgrader.py", line 24, in upgrade
>     _doValidation(deepcopy(profile), validator) # copy the profile, since validating mutates the object
>   File "config\profileUpgrader.py", line 57, in _doValidation
>     raise ValueError(errorString)
> ValueError: Unable to validate config file after upgrade: Key cursorBlinkRate : the value "-1" is too small.
> Full result: (value of false means the key was not present)
> [([u'braille'], u'cursorBlinkRate', VdtValueTooSmallError('the value "-1" is too small.',))]
> INFO - config.ConfigManager._loadConfig (03:03:45):
> Loading config: None
> INFO - config.profileUpgrader.upgrade (03:03:45):
> Checking schema version of config:
> {}
> INFO - core.main (03:03:45):
> NVDA version source-i6470-LimitCursorBlinkRate-076259c
> INFO - core.main (03:03:45):
> Using Windows version 6.2.9200 workstation
> INFO - core.main (03:03:45):
> Using Python version 2.7.11 (v2.7.11:6d1b6a68f775, Dec  5 2015, 20:32:19) [MSC v.1500 32 bit (Intel)]
> INFO - core.main (03:03:45):
> Using comtypes version 0.6.2
> INFO - synthDrivers.espeak.SynthDriver.__init__ (03:03:46):
> Using eSpeak version 1.48.15  16.Apr.15
> INFO - synthDriverHandler.setSynth (03:03:46):
> Loaded synthDriver espeak
> INFO - core.main (03:03:46):
> Using wx version 3.0.2.0 msw (classic)
> INFO - braille.initialize (03:03:46):
> Using liblouis version 3.0.0
> INFO - braille.BrailleHandler.setDisplayByName (03:03:46):
> Loaded braille display driver noBraille, current display has 0 cells.
> INFO - brailleInput.initialize (03:03:46):
> Braille input initialized
> WARNING - core.main (03:03:46):
> Java Access Bridge not available
> INFO - _UIAHandler.UIAHandler.MTAThreadFunc (03:03:46):
> UIAutomation: IUIAutomation3
> INFO - core.main (03:03:46):
> NVDA initialized
> INFO - config.ConfigManager.save (03:03:49):
> Base configuration saved
> INFO - core.main (03:03:55):
> Exiting
> INFO - config.ConfigManager.save (03:03:55):
> Base configuration saved
> INFO - __main__ (03:03:57):
> NVDA exit